contracts/examples/erc1155/src/lib.rs

```diff
@@ -0,0 +1,387 @@
+#![no_std]
+
+elrond_wasm::imports!();
+
+const INTERFACE_SIGNATURE_ERC165: u32 = 0x01ffc9a7;
+const INTERFACE_SIGNATURE_ERC1155: u32 = 0xd9b67a26;
+
+#[elrond_wasm_derive::contract(Erc1155Impl)]
+pub trait Erc1155 {
+	#[init]
+	fn init(&self) {}
+
+	// endpoints
+
+	/// `value` is amount for fungible, token_id for non-fungible
+	#[endpoint(safeTransferFrom)]
+	fn safe_transfer_from(
+		&self,
+		from: Address,
+		to: Address,
+		type_id: BigUint,
+		value: BigUint,
+		_data: &[u8],
+	) -> SCResult<()> {
+		let caller = self.get_caller();
+
+		require!(to != Address::zero(), "Can't transfer to address zero");
+		require!(self.is_valid_type_id(&type_id), "Toke id is invalid");
+		require!(
+			caller == from || self.get_is_approved(&caller, &from),
+			"Caller is not approved to transfer tokens from address"
+		);
+
+		if self.get_is_fungible(&type_id) == true {
+			let amount = &value;
+
+			require!(amount > &0, "Must transfer more than 0");
+			let balance = match self.get_balance_mapper(&from).get(&type_id) {
+				Some(b) => b,
+				None => return sc_error!("Address has no tokens of that type"),
+			};
+			require!(amount <= &balance, "Not enough balance for id");
+
+			self.decrease_balance(&from, &type_id, &amount);
+			self.increase_balance(&to, &type_id, &amount);
+		} else {
+			let token_id = &value;
+
+			require!(
+				self.is_valid_token_id(&type_id, &token_id),
+				"Token type-id pair is not valid"
+			);
+			require!(
+				self.get_token_owner(&type_id, &token_id) == from,
+				"_from_ is not the owner of the token"
+			);
+
+			let amount = BigUint::from(1u32);
+			self.decrease_balance(&from, &type_id, &amount);
+			self.increase_balance(&to, &type_id, &amount);
+
+			self.set_token_owner(&type_id, token_id, &to);
+		}
+
+		// self.transfer_single_event(&caller, &from, &to, &id, &amount);
+
+		if self.is_smart_contract_address(&to) {
+			// TODO: async-call
```

Comment:
in case of smart contracts you need to set an async call and have the state for increase / decrease in a pending state - thus if the async call fails you should revert those.